File: tests/matlab/test_spher.m
function tests = test_spher
% TEST_SPHER  Tests for spher.m (sphering matrix: 2*inv(sqrtm(cov(data')))).
% Properties: symmetric; whitens the channel covariance so that
% S*C*S' = 4*I (since the factor is 2). Single channel has a closed form.
tests = functiontests(localfunctions);
end

function testSingleChannelClosedForm(testCase)
% 1 channel: cov(data') = var(data) (N-1 normalized) = 2.5 -> 2/sqrt(2.5).
data = [1 2 3 4 5];
expected = 2 / sqrt(2.5);
verifyEqual(testCase, spher(data), expected, 'RelTol', 1e-10);
end

function testSymmetric(testCase)
% Sphering matrix of a symmetric PSD covariance is symmetric.
data = [2 1 0 1 3 2; 0 2 1 3 1 0; 1 0 3 1 2 1];
S = spher(data);
verifyLessThan(testCase, norm(S - S.'), 1e-9);
end

function testWhitensCovarianceToFourI(testCase)
% S = 2*inv(sqrtm(C)) -> S*C*S' = 4*I.
data = [2 1 0 1 3 2; 0 2 1 3 1 0; 1 0 3 1 2 1];
S = spher(data);
C = cov(data');
verifyLessThan(testCase, norm(S * C * S.' - 4 * eye(3)), 1e-6);
end

function testNoArgShowsHelp(testCase)
% nargin<1 branch: prints help and returns without error.
out = evalc('spher()');
verifyTrue(testCase, ~isempty(out));
end

File: tests/test_spher.py
"""Parity tests for spher - sphering matrix (EEGLAB spher.m).

``spher(data) = 2 * inv(sqrtm(cov(data.T)))``. The result is symmetric and
whitens the channel covariance so that ``S @ C @ S.T == 4*I`` (the factor is 2).
Single-channel data has the closed form ``2 / sqrt(var(data))``. Expectations
mirror tests/matlab/test_spher.m (confirmed on MATLAB R2025a).
"""

from __future__ import annotations

import numpy as np
import pytest

from eegprep.functions.sigprocfunc.spher import spher

pytestmark = pytest.mark.parity

# Deterministic full-rank 3-channel x 6-sample data (matches the MATLAB test).
DATA = np.array(
    [
        [2.0, 1.0, 0.0, 1.0, 3.0, 2.0],
        [0.0, 2.0, 1.0, 3.0, 1.0, 0.0],
        [1.0, 0.0, 3.0, 1.0, 2.0, 1.0],
    ]
)


def test_single_channel_closed_form():
    # 1 channel: cov(data.T) = var(data) (N-1) = 2.5 -> 2/sqrt(2.5).
    data = np.array([[1.0, 2.0, 3.0, 4.0, 5.0]])
    np.testing.assert_allclose(np.squeeze(spher(data)), 2 / np.sqrt(2.5), rtol=1e-10)


def test_symmetric():
    # Sphering matrix of a symmetric PSD covariance is symmetric.
    s = spher(DATA)
    np.testing.assert_allclose(s, s.T, atol=1e-9)


def test_whitens_covariance_to_four_i():
    # S = 2*inv(sqrtm(C)) -> S @ C @ S.T = 4*I.
    s = spher(DATA)
    c = np.cov(DATA)  # rowvar=True, ddof=1 == MATLAB cov(data.T)
    np.testing.assert_allclose(s @ c @ s.T, 4 * np.eye(3), atol=1e-6)
